### PR TITLE
♻️ 139 Move Specifications from Application to Domain Layer

### PR DESCRIPTION
**Pull request by danielmackay**
Reasoning provided via ADR: https://github.com/SSWConsulting/SSW.CleanArchitecture/blob/139-consider-location-of-specifications/docs/adr/20230810-move-specifications-to-the-domain.md

### CODE REVIEW

**Comment by danielmackay** [comment]
Am also considering removing `AllTodoItemSpec` as I don't see it adding any value.

File: docs/adr/20230810-move-specifications-to-the-domain.md [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+
+Chosen option: **Option 1 - Store Specfications in the Domain Layer**, as it allows us to move more business logic to the Domain Layer in a persistent ignorant way, and makes it easier to keep the specs up to date when the aggregate root changes.
+
+## Pros and Cons of the Options
+
+### Option 1 - Store Specfications in the Domain Layer
+
+- ✅ Move more business logic to the Domain Layer in a persistent ignorant way
+- ✅ Specs used to load entites are stored alongside the aggregate roots they are used to load.  This makes it easier to keep the specs up to date when the aggregate root changes.
+- ❌ Extra depdendency added to the Domain Layer
```
Comment (william-liebenberg) [inline docs/adr/20230810-move-specifications-to-the-domain.md:37]: which dependency?

File: src/Domain/TodoItems/AllTodoItemSpec.cs [hunk 2]
```diff
@@ -1,8 +1,9 @@
...
-namespace SSW.CleanArchitecture.Application.Features.TodoItems.Specifications;
-
-public sealed class AllTodoItemSpec : Specification<TodoItem>
-{
+using Ardalis.Specification;
+using SSW.CleanArchitecture.Domain.TodoItems;
+
+namespace SSW.CleanArchitecture.Application.Features.TodoItems.Specifications;
+
+// TODO DM: Consider removing this
```
Comment (william-liebenberg) [inline src/Domain/TodoItems/AllTodoItemSpec.cs:6]: remove why?

**Review by william-liebenberg — APPROVED** [review]
Looks good - small update to the ADR might make things a bit clearer.

File: src/Domain/TodoItems/AllTodoItemSpec.cs (hunk 2 above)
Comment (danielmackay) [reply src/Domain/TodoItems/AllTodoItemSpec.cs:6]: It's not actually encapsulating any query logic.

I'm also of the view that not all queries need to be implemented via specifications.  I would err on the siide of writing queries directly with EF, and refactoring to use specifications when the query needs to be shared/re-used.

File: docs/adr/20230810-move-specifications-to-the-domain.md (hunk 1 above)
Comment (danielmackay) [reply docs/adr/20230810-move-specifications-to-the-domain.md]: Have updated to reference `Ardalis.Specifications` depdenency.

File: src/Domain/TodoItems/AllTodoItemSpec.cs (hunk 2 above)
Comment (danielmackay) [reply src/Domain/TodoItems/AllTodoItemSpec.cs]: This has now been removed.

File: src/Domain/TodoItems/TodoItemByIdSpec.cs [hunk 3]
```diff
@@ -0,0 +1,12 @@
+﻿using Ardalis.Specification;
+using SSW.CleanArchitecture.Domain.TodoItems;
+
+namespace SSW.CleanArchitecture.Application.Features.TodoItems.Specifications;
```
Comment (Hona) [inline src/Domain/TodoItems/TodoItemByIdSpec.cs:4]: Looks like wrong namespace?

File: src/Domain/TodoItems/TodoItemByTitleSpec.cs [hunk 4]
```diff
@@ -1,12 +1,12 @@
...
-    public TodoItemByTitleSpec(string title)
-    {
-        Query.Where(i => i.Title == title);
-    }
-}
\ No newline at end of file
+using Ardalis.Specification;
+using SSW.CleanArchitecture.Domain.TodoItems;
+
+namespace SSW.CleanArchitecture.Application.Features.TodoItems.Specifications;
```
Comment (Hona) [inline src/Domain/TodoItems/TodoItemByTitleSpec.cs:4]: namespace

**Review by Hona — APPROVED** [review]
I like moving the Specs to domain, as its persistance unaware.

File: src/Domain/TodoItems/TodoItemByIdSpec.cs (hunk 3 above)
Comment (danielmackay) [reply src/Domain/TodoItems/TodoItemByIdSpec.cs:4]: Good catch.  Have fixed up.

File: src/Domain/TodoItems/TodoItemByTitleSpec.cs (hunk 4 above)
Comment (danielmackay) [reply src/Domain/TodoItems/TodoItemByTitleSpec.cs:4]: Good catch.  Have fixed up.

**Pull request merged**